### ISSUE TITLE
Qualifiers are not percent-encoded correctly

### ISSUE THREAD

**Issue by Shnatsel** [phylum-dev/purl#11]
# Overview

```rust
builder.with_qualifier("vcs_url", "git+https://github.com/rustsec/rustsec.git@6f82509cb43a");
``` 
ends up encoded as `vcs_url=git+https://github.com/rustsec/rustsec.git%406f82509cb43a` - note the non-escaped `+` symbol. 

By contrast, the official spec [provides an example](https://github.com/package-url/purl-spec/blob/07d2bdea6d9610b52772866c7ed7859e5566f318/PURL-TYPES.rst) where the `+` sign is percent-encoded: `pkg:generic/bitwarderl?vcs_url=git%2Bhttps://git.fsfe.org/dxtr/bitwarderl%40cc55108da32`

# Expected Behavior
The correct percent encoding character set for URL parameters being used, matching the URL and PURL specifications.

**Comment by Shnatsel** [comment]
https://datatracker.ietf.org/doc/html/rfc3986#section-2.2 specifies the characters to be escaped, including `+`.

https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#a-purl-is-a-url specifies that any PURL should be a compliant URL.

**Comment by Shnatsel** [comment]
Also, the spec states:

> A key must NOT be percent-encoded

While the code does, in fact, percent-encode the key:

https://github.com/phylum-dev/purl/blob/4d17bd88ed4156a2e17c8d8214c358be55f12863/purl/src/format.rs#L61-L68

**Comment by Shnatsel** [comment]
Actually, upon closer inspection of the WHATWG URL spec, the RFC 3986, and the PURL spec, I think the PURL spec examples are just nonsensical and contradict the spec.

I'll post and link a detailed write-up about this shortly.

**Comment by Shnatsel** [comment]
Okay, I've documented the contradictions in the spec: https://github.com/Shnatsel/cyclonedx-rust-cargo/blob/f79d239f8bf098de3666f271c55b47e631099d43/cargo-cyclonedx/src/urlencode.rs

I'm not really sure what the takeaway for this crate should be. I just added an extra layer of very aggressive percent encoding to deal with this in my code.

**Comment by kylewillmon** [comment]
Looks like we need to use the [application/x-www-form-urlencoded percent-encode set][1] rather than the query percent-encode set because PURL uses the `application/x-www-form-urlencoded` format by nature of using search parameters (see the [`URLSearchParams` interface][2] in the URL spec).

[1]: https://url.spec.whatwg.org/#application-x-www-form-urlencoded-percent-encode-set
[2]: https://url.spec.whatwg.org/#interface-urlsearchparams

**Comment by matt-phylum** [comment]
I believe the current behavior is correct. See package-url/purl-spec#39.

**Comment by matt-phylum** [comment]
In regards to the comment at the top of the cyclonedx-rust-cargo urlencode.rs file, I think the issue is that people would really like to compare PURLs as strings, which requires that every PURL has exactly one canonical encoding. This is only possible if the PURL spec is very exact about how characters are escaped and has a very thorough conformance test suite. This is far from the case, so PURL comparison needs to be more like [URL equivalence](https://url.spec.whatwg.org/#url-equivalence) where it can be compared as a string only if both strings were canonicalized and serialized by the same algorithm. Overescaping qualifier values should be okay, and software that works with PURLs should expect to encounter PURLs that are overescaped by people who use something like `encodeUriComponent`.

There's also a related problem that starts to come up when some PURLs for the same package contain optional qualifiers, which makes things significantly more complicated, especially because whether a qualifier is critical or not is not part of the qualifier. Even worse, if you're relying on the checksum in an ecosystem to distinguish different packages that have the same name and version, it's possible to have two PURLs with mutually exclusive sets of hash algorithms, in which case it's impossible to tell whether the two PURLs refer to the same package without having the package itself.

## About + signs

I think interpreting a '+' as a space is a significant decoding error. ['+' being a space is specifically a form-urlencoded behavior](https://url.spec.whatwg.org/#urlencoded-parsing), unrelated to percent encoding. The PURL spec defines something similar to form-urlencoded, but does not name form-urlencoded once, nor does it specify any special parsing rules for '+' signs. The examples and tests never show what a '+' in a qualifier means.

It looks like there's a problem with the swid package type documentation where spaces in _URL path components_ are replaced by '+' signs, which is definitely not right. Based on Adobe's documentation it should be `pkg:swid/Adobe%20Inc./Adobe%20InDesign%20CC@[some version]?tag_id=CreativeCloud-CS6-Win-GM-MUL`. In another example, a '+' sign in a URL fragment component means literally a '+', not a space.

This is probably the kind of problem that lead to the aggressive encoding used by form-urlencoded. If your PURL implementation percent encodes spaces and plus signs, it doesn't matter if the decoder is handling plus signs correctly because your PURLs don't contain them. I feel like encoding characters that do not require encoding is a deviation from the current PURL spec, but it should be safe as long as people aren't doing cross-implementation string comparisons, and I'm not totally opposed to doing it, especially for '+' signs, although it would result in an implementation that fails the test suite if the test suite covered this case.

## "A key must NOT be percent-encoded"

I think this is a problem with the spec. If the key contains characters that must be percent encoded, they must be percent encoded, or else the PURL will not parse correctly. The PURL spec as currently written talks about encoding the least number of characters possible to be valid (["must be encoded as […] _elsewhere_"](https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#character-encoding) and at least one of the tests in the test suite), in contrast to form-urlencoded which encodes as many characters as reasonably possible to avoid problems with broken parser implementations.

I think it's more accurate to say, if we're talking about encoding the minimum number of characters, that the qualifier key must not contain any characters that would require percent encoding, and this is true because

> The key must be composed only of ASCII letters and numbers, '.', '-' and '_' (period, dash and underscore)

However, if we're talking about encoding other characters that are not ambiguous, then it's possible, although unlikely, that it becomes necessary to specify something like like "The key must be comprised only of ASCII letters and numbers, '.', '-' and '_' (period, dash and underscore) _which do not require escaping and must not be unnecessarily escaped here_." However, even with form-urlencoded, this is almost exactly the set of characters that are never escaped. Somebody would need to encode everything that isn't ASCII alphanumeric or something like that for it to become relevant that the key is expected not to contain percent encoded characters.

**Comment by Shnatsel** [comment]
I see that the current behavior is compliant with the URL spec, and the PURL spec does not provide the necessary detail to determine if this behavior is correct or not. So I am going to go ahead and close this. Thank you for investigating this!

**Comment by kylewillmon** [comment]
I don't think we can dismiss this so easily... Here is a test against the [official JS implementation][js] of PURL:

[js]: https://github.com/package-url/packageurl-js

<details>
<summary>See JS code</summary>

```js
import { PackageURL } from 'packageurl-js';
const x = PackageURL.fromString('pkg:generic/bitwarderl?vcs_url=git+https://github.com/rustsec/rustsec.git%406f82509cb43a');
console.log(x.qualifiers.vcs_url);
```

</details>

The output is:

```
git https://github.com/rustsec/rustsec.git@6f82509cb43a
```

Similarly, here is a test against [the official Go package][go]:

[go]: https://github.com/package-url/packageurl-go

<details>
<summary>See Go code</summary>

```go
package main

import (
	"fmt"

	"github.com/package-url/packageurl-go"
)

func main() {
	instance, err := packageurl.FromString("pkg:generic/bitwarderl?vcs_url=git+https://github.com/rustsec/rustsec.git%406f82509cb43a")
	if err != nil {
		panic(err)
	}
	fmt.Printf("%s", instance.Qualifiers.Map()["vcs_url"])
}
```

</details>

This **also** outputs:

```
git https://github.com/rustsec/rustsec.git@6f82509cb43a                                                                                                                                                                                                                                                                         
```

This is either a bug in our code or it is a bug in both official implementations...

> The PURL spec defines something similar to form-urlencoded, but does not name form-urlencoded once, nor does it specify any special parsing rules for '+' signs. The examples and tests never show what a '+' in a qualifier means.

By getting close to form-urlencoded without explicitly saying that this **isn't** form-urlencoded, the spec has made itself ambiguous... And I think that leaves us with no choice but to percent encode `+` signs.

**Comment by matt-phylum** [comment]
Looking at the Javascript version I think there are multiple problems with the way things are being encoded and decoded. It does things like call `encodeURIComponent` and then string replace specific escape sequences back to their unencoded form. It uses form-urlencoded when decoding (by parsing the PURL as a URL), but not when encoding. For some reason, it throws an exception if your properly encoded version if it isn't encoded as if by `encodeURIComponent`, after it has successfully parsed the version.

The Go version, instead of encodeURIComponent, uses url.QueryEscape, which is supposed to be for form-urlencoded, "for compatibility with other implementations and the purl-spec". If it were incompatible with the spec, that wouldn't be a compatibility problem--it'd be wrong. However, after encoding it replaces '+' with '%20' because it is encoding every component of the PURL using the same function. It does decode '+' as ' ' in qualifiers, seemingly intentionally since it doesn't do that for other components, even if that's not part of the spec.

They're both nonconformant.

Both the Javascript version and the Go version fail the canonicalization part of the test suite from the spec repository because they escape characters that are not supposed to be escaped. The Javascript version includes a version of the test suite which does not contain PURLs which it fails to canonicalize. The Go version is missing the code to verify that it produces the expected strings.

Examples of both:

```
  6) PackageURL
       test-suite-data
         should be able to convert valid PackageURLs to a string:

      AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
+ actual - expected

+ 'pkg:huggingface/microsoft/deberta-v3-base@559062ad13d311b87b2c455e67dcd5f1c8f65111?repository_url=https%3A//hub-ci.huggingface.co'
- 'pkg:huggingface/microsoft/deberta-v3-base@559062ad13d311b87b2c455e67dcd5f1c8f65111?repository_url=https://hub-ci.huggingface.co'
      + expected - actual

      -pkg:huggingface/microsoft/deberta-v3-base@559062ad13d311b87b2c455e67dcd5f1c8f65111?repository_url=https%3A//hub-ci.huggingface.co
      +pkg:huggingface/microsoft/deberta-v3-base@559062ad13d311b87b2c455e67dcd5f1c8f65111?repository_url=https://hub-ci.huggingface.co
      
      at Context.<anonymous> (test/package-url.spec.js:138:18)
      at processImmediate (node:internal/timers:466:21)
```

```
--- FAIL: TestToStringExamples (0.00s)
    packageurl_test.go:217: docker uses qualifiers and hash image id as versions failed: pkg:docker/customer/dockerimage@sha256%3A244fd47e07d1004f0aed9c?repository_url=gcr.io != pkg:docker/customer/dockerimage@sha256:244fd47e07d1004f0aed9c?repository_url=gcr.io
```

The problem with spec ambiguity is that the (incomplete) test suite works by testing that either an input produces an error or exactly the expected output, down to the encoded characters, but then it doesn't give examples of all the characters in all the positions.

Maybe I should make a PR on the PURL spec repository specifying that it is not x-www-form-urlencoded but that '+' should be encoded as %2B in qualifiers to avoid problems with implementations that treat it as form-urlencoded, and adds tests to ensure that implementations encode '+' as %2B in qualifiers, encode '+' as '+' everywhere else, and never decode '+' as ' '. Making it form-urlencoded would make any existing conformant implementation nonconformant and make qualifiers more difficult for humans to read.

**Comment by matt-phylum** [comment]
I checked some of the other implementations under the package-url organization.

packageurl-java uses `URLEncoder.encode()`, which is for form-urlencoded, but then it changes '+' to %20 (good), '*' to %2A (why? This character is already supposed to be encoded if it's form-urlencoded, but is not supposed to be encoded according to PURL), and %7E to '~' (why?). This is a different encoding scheme than the other implementations. For decoding, it's broken and decodes '+' as ' ', but at least for all PURLs generated by packageurl-java the + character is (in violation of the spec) always encoded, regardless of position in the PURL. This means that packageurl-java does not pass the PURL test suite. It has the same problem as the Go implementation where the tests never verify that the PURL serializes into the expected string: https://github.com/package-url/packageurl-java/blob/bd3241aaa27ea297766a0a99e48b48ca17465956/src/test/java/com/github/packageurl/PackageURLTest.java#L294-L299

packageurl-php uses `rawurlencode` which "Encodes the given string according to RFC 3986", except that that description does not make sense because RFC 3986 does not define an exact algorithm for string encoding. Apparently from the return value, it's form-urlencoding with %20 instead of '+'. However, then packageurl-php unescapes %3A to ':' and '%2F' to '/'. The tests are implemented correctly, but use an old test suite file. Because it undecodes ':', it might actually pass the current test suite. '+' is decoded as '+'.

packageurl-python uses `urllib.parse.quote` for encoding, but strangely handles the utf-8 encoding itself and uses string replacement to unescape %3A as ':' instead of just telling `quote` not to escape that character in the first place. By default, `quote` doesn't escape '/' and the behavior should be similar to what packageurl-php does, so it might pass the test suite, which is correctly implemented to verify serialization. '+' is decoded as '+'.

packageurl-dotnet uses `WebUtility.UrlEncode`, which looks like a nonconformant implementation of form-urlencoded. Certain characters are decoded by string replacement, but which characters depends on the context. ' ' is incorrectly encoded as '+' as if it were form-urlencoding (even though it's not quite), even in places where that is not allowed, such as the package name. This implementation definitely fails the test suite. The existing tests only cover parsing.

packageurl-ruby is similarly broken. It uses `URI.encode_www_form_component`, which is form-urlencoded, and it uses it in all cases, including the package version. '+' is decoded as ' '. This implementation definitely fails the test suite. The repo does not contain the test suite or any code to run it.

packageurl.rs uses a [custom percent encode set](https://github.com/package-url/packageurl.rs/blob/eddaaedfe9541ac99367a5b17d8ba53edeaa8b52/src/purl.rs#L18-L37). '+' is '+'. It seems like it should pass the test suite, which it implements correctly for serialization, but it seems like it only runs a small subset of the tests. (this implementation, phylum-dev/purl, also doesn't run all the tests, because it doesn't support all the package types. I should add support for parsing and unparsing using strings instead of the package type enum to get better coverage).

packageurl-swift uses a [custom percent encode set](https://github.com/package-url/packageurl-swift/blob/b3d01733dd9d866386c8ae92c34bc34027953992/Sources/PackageURL/PackageURL.swift#L425-L427). '+' is '+'. It probably passes the test suite, but contains a subset of the test suite manually converted into test code.

I think that's '+' is '+' in 4/9 and '+' is ' ' in 5/9 of the implementations in the package-url organization (ie not including phylum-dev/purl).

**Comment by lfrancke** [comment]
Impressive research! Thank you.

**Comment by matt-phylum** [comment]
Related: package-url/packageurl-java#97. I'd guess multiple PURL implementations were being used and one of them output a version with an unescaped '+' and then the Java implementation converted that into a space (unambiguously wrong because the version is definitely not form-urlencoded), and then wrote it back out as a '+'.

**Comment by matt-phylum** [comment]
I've put together some code that allows me to uniformly exercise all nine of the parsers in the package-url organization and this one.

Ignoring overescaping in canonical PURLs, the PURL ecosystem has the following problems with spaces and plus signs:

# Plus signs outside qualifiers are converted into spaces (3/10)

packageurl-dotnet, packageurl-java, and packageurl-ruby incorrectly convert a PURL like `pkg:t/n+s/na+me@ve+r#s+p` to something like `pkg:t/n s/na me@ve r#s p`. This is definitely wrong. Nowhere does the PURL spec say to do this, and these parts of the PURL are definitely not form-urlencoded because they are not key value pairs.

# Plus signs inside qualifiers are converted into spaces (5/10)

packageurl-dotnet, packageurl-go, packageurl-java, packageurl-js, and packageurl-ruby incorrectly convert a PURL like `pkg:t/n?k=va+l` to something like `pkg:t/n?k=va l`. This is not part of the PURL spec, but could be implied because the qualifiers look a lot like they are form-urlencoded.

packageurl.rs, packageurl-php, packageurl-python, packageurl-swift and phylum-dev/purl do not perform this conversion.

# Spaces outside qualifiers are converted into plus signs (2/10)

packageurl-dotnet and packageurl-ruby incorrectly convert a PURL like `pkg:t/n s/na me@ve r#s p` to something like `pkg:t/n+s/na+me@ve+r#s+p`. This is the inverse of the previously mentioned problem and is definitely wrong for the same reasons.

This is not symmetrical to the inverse problem because packageurl-java escapes plus signs as %2B and spaces as %20.

# Spaces outside qualifiers are parse errors (2/10)

packageurl-java and packageurl-js unexpectedly fail to parse `pkg:t/n s/na me@ve r#s p`. This PURL is not canonical, but it's still valid and users may enter it.

# Spaces inside qualifiers are converted into plus signs (3/10)

packageurl-dotnet, packageurl-go, and packageurl-ruby incorrectly convert a PURL like `pkg:t/n?k=va l` into `pkg:t/n?k=va l`. This is the inverse of the previously mentioned problem and is wrong for the same reason.

packageurl.rs, packageurl-js, packageurl-php, packageurl-python, packageurl-swift, phylum-dev/purl do not perform this conversion. (packageurl-java doesn't either, but fails to parse the input)

This is not symmetrical to the inverse problem because packageurl-java fails and packageurl-js escapes plus signs as %2B and spaces as %20 (in qualifiers).

# Spaces inside qualifiers are a parse error (1/10)

packageurl-java unexpectedly fails to parse `pkg:t/n?k=va l`. This PURL is not canonical, but it's still valid and users may enter it.

**Comment by lfrancke** [comment]
I'll repeat myself: Impressive. Do you plan any followups here? As in: Raise issues in all those projects?

**Comment by Shnatsel** [comment]
I believe the first step is to add test cases to the specification repository.

**Comment by matt-phylum** [comment]
I created a PR on the spec repository specifying a behavior and adding tests for it: https://github.com/package-url/purl-spec/pull/261

Making it x-www-form-urlencoded wouldn't be terrible either, but it can't be both at the same time like it is in the current set of implementations.

**Comment by Shnatsel** [comment]
I believe this can be closed now. The immediate issue is fixed in v0.1.3. 

There is still the upstream specification interoperability PR, but it has stalled, and that's no longer an issue with this crate.
